<a href="https://colab.research.google.com/github/Alejods90/Challenge-Telecom-1/blob/main/Challenge_Telecom_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# New Section

In [ ]:
import pandas as pd

pd.read_json('/content/drive/MyDrive/TelecomX_Data.json')

customerID Churn                                           customer  \
0     0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1     0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2     0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3     0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4     0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
...          ...   ...                                                ...   
7262  9987-LUTYD    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
7263  9992-RRAMN   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7264  9992-UJOEL    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7265  9993-LHIEB    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7266  9995-HOTOH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                                  phone  \
0        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1       {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
...                                                 ...   
7262     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7263    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
7264     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7265     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7266  {'PhoneService': 'No', 'MultipleLines': 'No ph...   

                                               internet  \
0     {'InternetService': 'DSL', 'OnlineSecurity': '...   
1     {'InternetService': 'DSL', 'OnlineSecurity': '...   
2     {'InternetService': 'Fiber optic', 'OnlineSecu...   
3     {'InternetService': 'Fiber optic', 'OnlineSecu...   
4     {'InternetService': 'Fiber optic', 'OnlineSecu...   
...                                                 ...   
7262  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7263  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7264  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7265  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7266  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                                account  
0     {'Contract': 'One year', 'PaperlessBilling': '...  
1     {'Contract': 'Month-to-month', 'PaperlessBilli...  
2     {'Contract': 'Month-to-month', 'PaperlessBilli...  
3     {'Contract': 'Month-to-month', 'PaperlessBilli...  
4     {'Contract': 'Month-to-month', 'PaperlessBilli...  
...                                                 ...  
7262  {'Contract': 'One year', 'PaperlessBilling': '...  
7263  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7264  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7265  {'Contract': 'Two year', 'PaperlessBilling': '...  
7266  {'Contract': 'Two year', 'PaperlessBilling': '...  

[7267 rows x 6 columns]

In [ ]:
pd.json_normalize('/content/drive/MyDrive/TelecomX_Data.json')

NotImplementedError: 

In [ ]:
import json

#Leyendo el archivo json usando el módulo Python JSON
with open('/content/drive/MyDrive/TelecomX_Data.json','r') as f:
    datos = json.loads(f.read())

#Normalizando los datos con los parámetros record_path y meta
df = pd.json_normalize(datos, meta=['customer', 'phone', 'internet', 'account'])
df

customerID Churn customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO    No          Female                       0   
1     0003-MKNFE    No            Male                       0   
2     0004-TLHLJ   Yes            Male                       0   
3     0011-IGKFF   Yes            Male                       1   
4     0013-EXCHZ   Yes          Female                       1   
...          ...   ...             ...                     ...   
7262  9987-LUTYD    No          Female                       0   
7263  9992-RRAMN   Yes            Male                       0   
7264  9992-UJOEL    No            Male                       0   
7265  9993-LHIEB    No            Male                       0   
7266  9995-HOTOH    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
0                 Yes                 Yes                9                Yes   
1                  No                  No                9                Yes   
2                  No                  No                4                Yes   
3                 Yes                  No               13                Yes   
4                 Yes                  No                3                Yes   
...               ...                 ...              ...                ...   
7262               No                  No               13                Yes   
7263              Yes                  No               22                Yes   
7264               No                  No                2                Yes   
7265              Yes                 Yes               67                Yes   
7266              Yes                 Yes               63                 No   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
0                     No                      DSL  ...                   Yes   
1                    Yes                      DSL  ...                    No   
2                     No              Fiber optic  ...                    No   
3                     No              Fiber optic  ...                   Yes   
4                     No              Fiber optic  ...                    No   
...                  ...                      ...  ...                   ...   
7262                  No                      DSL  ...                    No   
7263                 Yes              Fiber optic  ...                    No   
7264                  No                      DSL  ...                   Yes   
7265                  No                      DSL  ...                    No   
7266    No phone service                      DSL  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                           No                  Yes                  Yes   
1                           No                   No                   No   
2                          Yes                   No                   No   
3                          Yes                   No                  Yes   
4                           No                  Yes                  Yes   
...                        ...                  ...                  ...   
7262                        No                  Yes                   No   
7263                        No                   No                   No   
7264                        No                   No                   No   
7265                       Yes                  Yes                   No   
7266                       Yes                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                          No         One year                      Yes   
1                         Yes   Month-to-month                       No   
2                          No   Month-to-month                      Yes   
3                         Yes   Month-to-month                    

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer.gender            7267 non-null   object 
 3   customer.SeniorCitizen     7267 non-null   int64  
 4   customer.Partner           7267 non-null   object 
 5   customer.Dependents        7267 non-null   object 
 6   customer.tenure            7267 non-null   int64  
 7   phone.PhoneService         7267 non-null   object 
 8   phone.MultipleLines        7267 non-null   object 
 9   internet.InternetService   7267 non-null   object 
 10  internet.OnlineSecurity    7267 non-null   object 
 11  internet.OnlineBackup      7267 non-null   object 
 12  internet.DeviceProtection  7267 non-null   object 
 13  internet.TechSupport       7267 non-null   objec

In [ ]:
df.dtypes

customerID                    object
Churn                         object
customer.gender               object
customer.SeniorCitizen         int64
customer.Partner              object
customer.Dependents           object
customer.tenure                int64
phone.PhoneService            object
phone.MultipleLines           object
internet.InternetService      object
internet.OnlineSecurity       object
internet.OnlineBackup         object
internet.DeviceProtection     object
internet.TechSupport          object
internet.StreamingTV          object
internet.StreamingMovies      object
account.Contract              object
account.PaperlessBilling      object
account.PaymentMethod         object
account.Charges.Monthly      float64
account.Charges.Total         object
dtype: object

In [ ]:
df['Churn'].unique()

array(['No', 'Yes', ''], dtype=object)

In [ ]:
df.isna().sum()

customerID                   0
Churn                        0
customer.gender              0
customer.SeniorCitizen       0
customer.Partner             0
customer.Dependents          0
customer.tenure              0
phone.PhoneService           0
phone.MultipleLines          0
internet.InternetService     0
internet.OnlineSecurity      0
internet.OnlineBackup        0
internet.DeviceProtection    0
internet.TechSupport         0
internet.StreamingTV         0
internet.StreamingMovies     0
account.Contract             0
account.PaperlessBilling     0
account.PaymentMethod        0
account.Charges.Monthly      0
account.Charges.Total        0
dtype: int64

In [ ]:
df.apply(lambda x: x.astype(str).str.strip() == '').sum()

customerID                     0
Churn                        224
customer.gender                0
customer.SeniorCitizen         0
customer.Partner               0
customer.Dependents            0
customer.tenure                0
phone.PhoneService             0
phone.MultipleLines            0
internet.InternetService       0
internet.OnlineSecurity        0
internet.OnlineBackup          0
internet.DeviceProtection      0
internet.TechSupport           0
internet.StreamingTV           0
internet.StreamingMovies       0
account.Contract               0
account.PaperlessBilling       0
account.PaymentMethod          0
account.Charges.Monthly        0
account.Charges.Total         11
dtype: int64

In [ ]:
df = df[df['Churn']!='']

In [ ]:
df

customerID Churn customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO    No          Female                       0   
1     0003-MKNFE    No            Male                       0   
2     0004-TLHLJ   Yes            Male                       0   
3     0011-IGKFF   Yes            Male                       1   
4     0013-EXCHZ   Yes          Female                       1   
...          ...   ...             ...                     ...   
7262  9987-LUTYD    No          Female                       0   
7263  9992-RRAMN   Yes            Male                       0   
7264  9992-UJOEL    No            Male                       0   
7265  9993-LHIEB    No            Male                       0   
7266  9995-HOTOH    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
0                 Yes                 Yes                9                Yes   
1                  No                  No                9                Yes   
2                  No                  No                4                Yes   
3                 Yes                  No               13                Yes   
4                 Yes                  No                3                Yes   
...               ...                 ...              ...                ...   
7262               No                  No               13                Yes   
7263              Yes                  No               22                Yes   
7264               No                  No                2                Yes   
7265              Yes                 Yes               67                Yes   
7266              Yes                 Yes               63                 No   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
0                     No                      DSL  ...                   Yes   
1                    Yes                      DSL  ...                    No   
2                     No              Fiber optic  ...                    No   
3                     No              Fiber optic  ...                   Yes   
4                     No              Fiber optic  ...                    No   
...                  ...                      ...  ...                   ...   
7262                  No                      DSL  ...                    No   
7263                 Yes              Fiber optic  ...                    No   
7264                  No                      DSL  ...                   Yes   
7265                  No                      DSL  ...                    No   
7266    No phone service                      DSL  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                           No                  Yes                  Yes   
1                           No                   No                   No   
2                          Yes                   No                   No   
3                          Yes                   No                  Yes   
4                           No                  Yes                  Yes   
...                        ...                  ...                  ...   
7262                        No                  Yes                   No   
7263                        No                   No                   No   
7264                        No                   No                   No   
7265                       Yes                  Yes                   No   
7266                       Yes                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                          No         One year                      Yes   
1                         Yes   Month-to-month                       No   
2                          No   Month-to-month                      Yes   
3                         Yes   Month-to-month                    

In [ ]:
df['account.Charges.Monthly']

0       65.60
1       59.90
2       73.90
3       98.00
4       83.90
        ...  
7262    55.15
7263    85.10
7264    50.30
7265    67.85
7266    59.00
Name: account.Charges.Monthly, Length: 7043, dtype: float64

In [ ]:
df['Charges_daily'] = df['account.Charges.Monthly']/30
display(df['Charges_daily'])

/tmp/ipython-input-3178898662.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Charges_daily'] = df['account.Charges.Monthly']/30


0                2.186667
1                1.996667
2                2.463333
3                3.266667
4                2.796667
                   ...   
7263             2.836667
7264             1.676667
7265             2.261667
7266             1.966667
Charges_daily         NaN
Name: Charges_daily, Length: 7044, dtype: float64

In [ ]:
df = df.drop('Charges_daily', errors='ignore')
df

customerID Churn customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO    No          Female                     0.0   
1     0003-MKNFE    No            Male                     0.0   
2     0004-TLHLJ   Yes            Male                     0.0   
3     0011-IGKFF   Yes            Male                     1.0   
4     0013-EXCHZ   Yes          Female                     1.0   
...          ...   ...             ...                     ...   
7262  9987-LUTYD    No          Female                     0.0   
7263  9992-RRAMN   Yes            Male                     0.0   
7264  9992-UJOEL    No            Male                     0.0   
7265  9993-LHIEB    No            Male                     0.0   
7266  9995-HOTOH    No            Male                     0.0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
0                 Yes                 Yes              9.0                Yes   
1                  No                  No              9.0                Yes   
2                  No                  No              4.0                Yes   
3                 Yes                  No             13.0                Yes   
4                 Yes                  No              3.0                Yes   
...               ...                 ...              ...                ...   
7262               No                  No             13.0                Yes   
7263              Yes                  No             22.0                Yes   
7264               No                  No              2.0                Yes   
7265              Yes                 Yes             67.0                Yes   
7266              Yes                 Yes             63.0                 No   

     phone.MultipleLines internet.InternetService  ...  \
0                     No                      DSL  ...   
1                    Yes                      DSL  ...   
2                     No              Fiber optic  ...   
3                     No              Fiber optic  ...   
4                     No              Fiber optic  ...   
...                  ...                      ...  ...   
7262                  No                      DSL  ...   
7263                 Yes              Fiber optic  ...   
7264                  No                      DSL  ...   
7265                  No                      DSL  ...   
7266    No phone service                      DSL  ...   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                           No                  Yes                  Yes   
1                           No                   No                   No   
2                          Yes                   No                   No   
3                          Yes                   No                  Yes   
4                           No                  Yes                  Yes   
...                        ...                  ...                  ...   
7262                        No                  Yes                   No   
7263                        No                   No                   No   
7264                        No                   No                   No   
7265                       Yes                  Yes                   No   
7266                       Yes                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                          No         One year                      Yes   
1                         Yes   Month-to-month                       No   
2                          No   Month-to-month                      Yes   
3                         Yes   Month-to-month                      Yes   
4                          No   Month-to-month                      Yes   
...                       ...              ...                      ...   
7262                       No         One year                       No   
7263                      Yes 

In [ ]:
# Create a copy of the DataFrame after filtering to avoid SettingWithCopyWarning
df = df.copy()

new_columns = {}
for col in df.columns:
    if '.' in col:
        parts = col.split('.')
        if len(parts) > 1:
            new_name = parts[-1].capitalize()
            new_columns[col] = new_name

df.rename(columns=new_columns, inplace=True)
display(df.head())

customerID Churn  Gender  Seniorcitizen Partner Dependents  Tenure  \
0  0002-ORFBO    No  Female            0.0     Yes        Yes     9.0   
1  0003-MKNFE    No    Male            0.0      No         No     9.0   
2  0004-TLHLJ   Yes    Male            0.0      No         No     4.0   
3  0011-IGKFF   Yes    Male            1.0     Yes         No    13.0   
4  0013-EXCHZ   Yes  Female            1.0     Yes         No     3.0   

  Phoneservice Multiplelines Internetservice  ... Deviceprotection  \
0          Yes            No             DSL  ...               No   
1          Yes           Yes             DSL  ...               No   
2          Yes            No     Fiber optic  ...              Yes   
3          Yes            No     Fiber optic  ...              Yes   
4          Yes            No     Fiber optic  ...               No   

  Techsupport Streamingtv Streamingmovies        Contract Paperlessbilling  \
0         Yes         Yes              No        One year              Yes   
1          No          No             Yes  Month-to-month               No   
2          No          No              No  Month-to-month              Yes   
3          No         Yes             Yes  Month-to-month              Yes   
4         Yes         Yes              No  Month-to-month              Yes   

      Paymentmethod Monthly    Total  Charges_daily  
0      Mailed check    65.6    593.3       2.186667  
1      Mailed check    59.9    542.4       1.996667  
2  Electronic check    73.9   280.85       2.463333  
3  Electronic check    98.0  1237.85       3.266667  
4      Mailed check    83.9    267.4       2.796667  

[5 rows x 22 columns]

In [ ]:
for col in df.columns:
    if df[col].dtype == 'object':
        # Check if the column contains only 'Yes' and 'No' or values that can be coerced to numeric
        if df[col].isin(['Yes', 'No']).all():
            df[col] = df[col].replace({'Yes': 1, 'No': 0}).astype(int)
        # Handle other object columns that might contain numeric-like strings
        elif df[col].str.isnumeric().all():
             df[col] = pd.to_numeric(df[col])


display(df.head())

/tmp/ipython-input-2190086144.py:5: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df[col] = df[col].replace({'Yes': 1, 'No': 0}).astype(int)
/tmp/ipython-input-2190086144.py:5: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df[col] = df[col].replace({'Yes': 1, 'No': 0}).astype(int)
/tmp/ipython-input-2190086144.py:5: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, 

customerID  Churn  Gender  Seniorcitizen  Partner  Dependents  Tenure  \
0  0002-ORFBO      0  Female            0.0        1           1     9.0   
1  0003-MKNFE      0    Male            0.0        0           0     9.0   
2  0004-TLHLJ      1    Male            0.0        0           0     4.0   
3  0011-IGKFF      1    Male            1.0        1           0    13.0   
4  0013-EXCHZ      1  Female            1.0        1           0     3.0   

   Phoneservice Multiplelines Internetservice  ... Deviceprotection  \
0             1            No             DSL  ...               No   
1             1           Yes             DSL  ...               No   
2             1            No     Fiber optic  ...              Yes   
3             1            No     Fiber optic  ...              Yes   
4             1            No     Fiber optic  ...               No   

  Techsupport Streamingtv Streamingmovies        Contract Paperlessbilling  \
0         Yes         Yes              No        One year                1   
1          No          No             Yes  Month-to-month                0   
2          No          No              No  Month-to-month                1   
3          No         Yes             Yes  Month-to-month                1   
4         Yes         Yes              No  Month-to-month                1   

      Paymentmethod  Monthly    Total  Charges_daily  
0      Mailed check     65.6    593.3       2.186667  
1      Mailed check     59.9    542.4       1.996667  
2  Electronic check     73.9   280.85       2.463333  
3  Electronic check     98.0  1237.85       3.266667  
4      Mailed check     83.9    267.4       2.796667  

[5 rows x 22 columns]

In [ ]:
df['Total'] = pd.to_numeric(df['Total'], errors='coerce')
display(df.info())

<class 'pandas.core.frame.DataFrame'>
Index: 7043 entries, 0 to 7266
Data columns (total 22 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   Churn             7043 non-null   int64  
 2   Gender            7043 non-null   object 
 3   Seniorcitizen     7043 non-null   float64
 4   Partner           7043 non-null   int64  
 5   Dependents        7043 non-null   int64  
 6   Tenure            7043 non-null   float64
 7   Phoneservice      7043 non-null   int64  
 8   Multiplelines     7043 non-null   object 
 9   Internetservice   7043 non-null   object 
 10  Onlinesecurity    7043 non-null   object 
 11  Onlinebackup      7043 non-null   object 
 12  Deviceprotection  7043 non-null   object 
 13  Techsupport       7043 non-null   object 
 14  Streamingtv       7043 non-null   object 
 15  Streamingmovies   7043 non-null   object 
 16  Contract          7043 non-null   object 
 17  

None

In [ ]:
df.describe()

Churn  Seniorcitizen      Partner   Dependents       Tenure  \
count  7043.000000    7043.000000  7043.000000  7043.000000  7043.000000   
mean      0.265370       0.162147     0.483033     0.299588    32.371149   
std       0.441561       0.368612     0.499748     0.458110    24.559481   
min       0.000000       0.000000     0.000000     0.000000     0.000000   
25%       0.000000       0.000000     0.000000     0.000000     9.000000   
50%       0.000000       0.000000     0.000000     0.000000    29.000000   
75%       1.000000       0.000000     1.000000     1.000000    55.000000   
max       1.000000       1.000000     1.000000     1.000000    72.000000   

       Phoneservice  Paperlessbilling      Monthly        Total  Charges_daily  
count   7043.000000       7043.000000  7043.000000  7032.000000    7043.000000  
mean       0.903166          0.592219    64.761692  2283.300441       2.158723  
std        0.295752          0.491457    30.090047  2266.771362       1.003002  
min        0.000000          0.000000    18.250000    18.800000       0.608333  
25%        1.000000          0.000000    35.500000   401.450000       1.183333  
50%        1.000000          1.000000    70.350000  1397.475000       2.345000  
75%        1.000000          1.000000    89.850000  3794.737500       2.995000  
max        1.000000          1.000000   118.750000  8684.800000       3.958333

In [ ]:
import plotly.express as px

In [ ]:
churn_counts = df['Churn'].value_counts().reset_index()
churn_counts.columns = ['Churn', 'Count']
fig = px.pie(churn_counts, values='Count', names='Churn', title='Churn Distribution')
fig.show()

# Task
Analyze the relationship between churn and other variables in the dataset and visualize the findings.

## Analyze categorical features

### Subtask:
Create visualizations (e.g., bar plots) to see how churn relates to categorical features like 'Gender', 'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', and 'PaymentMethod'.


**Reasoning**:
Create bar plots for each categorical feature to visualize the relationship with churn.



In [ ]:
categorical_cols = ['Gender', 'Partner', 'Dependents', 'Phoneservice', 'Multiplelines', 'Internetservice', 'Onlinesecurity', 'Onlinebackup', 'Deviceprotection', 'Techsupport', 'Streamingtv', 'Streamingmovies', 'Contract', 'Paperlessbilling', 'Paymentmethod']

for col in categorical_cols:
    churn_by_cat = df.groupby([col, 'Churn']).size().reset_index(name='Count')
    # Convert Churn to string for discrete colors in legend
    churn_by_cat['Churn'] = churn_by_cat['Churn'].astype(str)
    fig = px.bar(churn_by_cat, x=col, y='Count', color='Churn', title=f'Churn Distribution by {col}', color_discrete_sequence=px.colors.qualitative.Plotly)
    fig.show()

## Analizar Aspectos Numéricos




Crear diagramas de caja para visualizar la relación entre la cancelación de clientes y las características numéricas 'Tenure', 'Monthly' y 'Total'.


In [ ]:
numerical_cols = ['Tenure', 'Monthly', 'Total']

for col in numerical_cols:
    fig = px.box(df, y=col, x='Churn', title=f'Churn Distribution by {col}')
    fig.show()

Vamos a visualizar la relación entre 'SeniorCitizen' y 'Churn' utilizando gráficos apropiados.


In [ ]:
churn_by_seniorcitizen = df.groupby(['Seniorcitizen', 'Churn']).size().reset_index(name='Count')
display(churn_by_seniorcitizen)

Seniorcitizen  Churn  Count
0            0.0      0   4508
1            0.0      1   1393
2            1.0      0    666
3            1.0      1    476

**Reasoning**:
Create a bar chart to visualize the relationship between 'Seniorcitizen' and 'Churn' using the grouped data.



In [ ]:
# Convert Churn to string for discrete colors in legend
churn_by_seniorcitizen['Churn'] = churn_by_seniorcitizen['Churn'].astype(str)
fig = px.bar(churn_by_seniorcitizen, x='Seniorcitizen', y='Count', color='Churn', title='Churn Distribution by Senior Citizen Status', color_discrete_sequence=px.colors.qualitative.Plotly)
fig.show()

## Resumen:

### Hallazgos Clave del Análisis de Datos

* Entre las características categóricas, las siguientes muestran diferencias notables en la distribución de la **cancelación de clientes** (churn):
    * Los clientes con **'Servicio de Internet'** diferente de 'No' (DSL o Fibra óptica) tienen una tasa de cancelación más alta.
    * Los clientes sin **'Seguridad en Línea'**, **'Copia de Seguridad en Línea'**, **'Protección de Dispositivos'** o **'Soporte Técnico'** tienden a cancelar más.
    * Los clientes con un contrato **'Mes a mes'** cancelan significativamente más que aquellos con contratos a más largo plazo.
    * Los clientes que usan **'Facturación sin Papel'** tienen una tasa de cancelación más alta.
    * El **'Método de Pago'** también muestra variaciones en las tasas de cancelación entre los diferentes métodos.
* Para las características numéricas:
    * Los clientes que cancelaron ('Sí') tienen una **'Antigüedad'** menor en comparación con aquellos que no cancelaron ('No').
    * Los clientes que cancelaron ('Sí') tienden a tener cargos **'Mensuales'** más altos.
    * Los clientes que no cancelaron ('No') generalmente tienen cargos **'Totales'** más altos, probablemente debido a una mayor antigüedad.
* Con respecto a la característica **'Adulto Mayor'**:
    * No adultos mayores: 4508 no cancelaron, 1393 cancelaron.
    * Adultos mayores: 666 no cancelaron, 476 cancelaron.
    * Los adultos mayores tienen una tasa de cancelación más alta (aproximadamente 41.6%) en comparación con los no adultos mayores (aproximadamente 23.6%).

---

### Perspectivas o Próximos Pasos

* Se justifica una investigación más a fondo de las razones específicas detrás de la mayor cancelación entre los clientes con internet de fibra óptica, contratos mes a mes y facturación sin papel.
* Considerar la construcción de un modelo predictivo utilizando las características identificadas para predecir la cancelación de clientes e identificar a los clientes de alto riesgo.
